File: apps/api/services/upload_service.py
import requests
import json
import os

from apps.api.models import (
    Artist,
    Image,
    Album
)
from apps.extensions import db
from apps.api.utils import (
    ConflictError,
    path_to_images
)
from .genre_service import get_or_create_genre
from .artist_service import get_artists_ids

import pandas as pd


def get_artist_names(file_path):
    df = pd.read_csv(file_path)
    df = df.drop(['facebook', 'twitter', 'genre', 'mtv', 'website'], axis=1)
    df = df.dropna() # drop missing values
    artists = df.name.to_list()
    return artists


def validate_artist(resp_content):

    if not isinstance(resp_content, list):
        return False
    resp_content = resp_content[0]

    if 'strArtist' not in resp_content or 'idArtist' not in resp_content:
        return False

    if Artist.query.filter_by(id=resp_content.get('idArtist')).first() is not None:
        return False

    validate_field(resp_content, 'strCountryCode')
    validate_field(resp_content, 'strBiographyEN')
    validate_field(resp_content, 'intFormedYear')
    validate_field(resp_content, 'intMembers')
    validate_field(resp_content, 'strLabel')
    validate_field(resp_content, 'strWebsite')
    validate_field(resp_content, 'strFacebook')

    if "strStyle" not in resp_content or resp_content["strStyle"] is None:
        resp_content["strStyle"] = "UNKNOWN"

    return True


def validate_field(resp_content, field_name):
    if field_name not in resp_content:
        resp_content[field_name] = None


def upload_artists():
    """
    Gets the artist names from 10000-MTV-Music-Artists-page-1/2/3.csv
    Gets the artist details via www.theaudiodb.com/api
    Adds artist details with image, and genre to db
    :return:
    """
    file_paths = []
    for i in range(1, 4):
        file_paths.append(f'D:\\Coding\\Licenta\\10000-MTV-Music-Artists-page-{i}.csv')

    for file_index, file_path in enumerate(file_paths):
        artists = get_artist_names(file_path)

        for artist_index, artist in enumerate(artists):
            artist = artist[1:-1]
            req_link = 'https://www.theaudiodb.com/api/v1/json/1/search.php?s=' + artist
            resp = requests.get(req_link)

            try:
                resp_content = json.loads(resp.content.decode("UTF-8"))['artists']

                if not validate_artist(resp_content):
                    continue
                resp_content = resp_content[0]
            except Exception as e:
                print(e)
                raise ConflictError("Response content is not valid")

            add_new_artist(artist, artist_index, file_index, resp_content)


def add_new_artist(artist, artist_index, file_index, resp_content):
    """
    checks if artist can have an image attached,
    adds an image to db if so,
    maps genre to an object,
    adds artist to db
    :param artist: artist name
    :param artist_index:
    :param file_index:
    :param resp_content: response from audio db
    :return:
    """
    image_obj = None

    if resp_content["strArtistThumb"] is not None:
        artist_image_link = resp_content["strArtistThumb"]

        try:
            resp = requests.get(artist_image_link)
            image_name = f'artist_{resp_content.get("idArtist")}.jpg'
            image_path = os.path.join(path_to_images, image_name)

            with open(image_path, 'wb') as f:
                f.write(resp.content)
                image_obj = Image(path=image_path)
                db.session.add(image_obj)
                db.session.commit()

        except Exception as e:
            print(e)
            image_obj = None
            print('image not saved')

    artist_genre = get_or_create_genre(resp_content["strStyle"])

    artist_dict = {
        "name": resp_content['strArtist'],
        "origin_country": resp_content['strCountryCode'],
        "description": resp_content['strBiographyEN'],
        "genre": artist_genre.id,
        "website": resp_content['strWebsite'],
        "facebook_link": resp_content['strFacebook'],
        "members": resp_content['intMembers'],
        "id": resp_content['idArtist']
    }

    if image_obj is not None:
        artist_dict['image_id'] = image_obj.id
    try:
        artist_obj = Artist(**artist_dict)
        db.session.add(artist_obj)
        db.session.commit()
        print('Successfully added', artist)
    except Exception as e:
        print(e)


def upload_albums():
    artist_ids = get_artists_ids()

    for artist_id in artist_ids:
        link = f'https://theaudiodb.com/api/v1/json/1/album.php?i={artist_id}'
        albums = requests.get(link).json()['album']
        if albums is None:
            continue
        for album in albums:
            if validate_album(album) is False:
                continue
            image_obj = None
            if album.get('strAlbumThumb'):
                image_path = os.path.join(path_to_images, f"album_{album.get('idAlbum')}.jpg")
                resp_image = requests.get(album.get('strAlbumThumb'))

                with open(image_path, 'wb') as f:
                    f.write(resp_image.content)
                    image_obj = Image(path=image_path)
                    db.session.add(image_obj)
                    db.session.commit()
            if album.get("strStyle") is None:
                album["strStyle"] = "UNKNOWN"
            album_genre = get_or_create_genre(album.get("strStyle"))
            album_dict = {
                "id": album.get('idAlbum'),
                "artist_id": artist_id,
                "name": album.get('strAlbum'),
                "description": album.get('strDescriptionEN'),
                "image_id": image_obj.id if image_obj is not None else None,
                "release_year": album.get('intYearReleased'),
                "genre": album_genre.id
            }

            try:
                album_obj = Album(**album_dict)
                db.session.add(album_obj)
                db.session.commit()
                print('Successfully added', album_obj.name)
            except Exception as e:
                print(e)
                raise ConflictError("Error while saving the album")


def validate_album(album):

    if Album.query.filter_by(id=album.get('idAlbum')).first() is not None:
        return False
    if album.get('strReleaseFormat').lower() != 'album':
        return False
    if not isinstance(album.get('strAlbum'), str):
        return False
    return True

File: node-backend/app/service/upload.service.js
const path = require('path');
const csv = require('csv-parse');
const axios = require('axios');
const fs = require('fs');
const https = require('https');
const Genre = require('../models/genre.model')
const Artist = require('../models/artist.model');
const Album = require('../models/album.model');
const sharp = require('sharp')


const uploadArtists = async () => {
  const filePath = path.join(process.cwd(), '10000-MTV-Music-Artists-page-1.csv');
  const artists = [];
  fs.createReadStream(filePath)
    .pipe(csv())
    .on('data', (row) => {
      artists.push(row);
    })
    .on('end', () => {
      for(const [index, value] of artists.entries()){
        if (index === 60) break;
        const artistName = value[0].substring(1, value[0].length - 1)
        const url = 'theaudiodb.com/api/v1/json/1/search.php?s='.concat(artistName)
        axios.get(url).then(async resp =>  {
          if (resp.data.artists != null) {
            const artist = resp.data.artists[0];
            if (await validatePayload(artist)) {
              const artistImage = await extractArtistImage(artist);
              const artistGenre = await getOrCreateGenre(artist.strGenre);
              const data = {
                'id': artist.idArtist,
                'name': artist.strArtist,
                'originCountry': artist.strCountryCode,
                'description': artist.strBiographyEN,
                'genreId': artistGenre.dataValues.id,
                'website': artist.strWebsite,
                'facebookLink': artist.strFacebook,
                'members': artist.intMembers || 1,
                'formedYear': artist.intFormedYear,
                'recordLabel': artist.strLabel,
                'image': artistImage
              }
              const artistObj = await Artist.build(data)
              await artistObj.save()
              // add all the albums for artist
              await uploadAlbumsForArtist(artistObj.id);
            }
          }
        })
      .catch ((e) => {
        console.log(e);
      })}
    });
}

const uploadAlbumsForArtist = async (artistId) => {
  const url = 'https://theaudiodb.com/api/v1/json/1/album.php?i='.concat(artistId);
  const resp = await axios.get(url);

  for(const item in resp.data){
    if (item.strReleaseFormat !== 'Album') continue;

    const albumGenre = await getOrCreateGenre(item.strGenre);
    const albumImage = await extractAlbumImage(item);
    const data = {
      "id": item.idAlbum,
      "name": item.strAlbum,
      "artistId": item.idArtist,
      "description": item.strDescriptionEN,
      "image": albumImage,
      "genre": albumGenre,
      "releaseYear": item.intYearReleased
    }
    const albumObj = await Album.build(data);
    await albumObj.save();
  }
}

async function extracted(url, imageName) {
  try {
    const pathToImages = path.join(process.cwd(), 'images');

    if (!fs.existsSync(pathToImages)) {
      fs.mkdirSync(pathToImages, {recursive: true});
    }

    const imagePath = path.join(pathToImages, imageName);
    await axios.get(url, {
      responseType: 'arraybuffer'
    })
      .then(response => sharp(response.data).resize(350, 350).toFile(imagePath));

    return imagePath;
  } catch (e) {
    console.log(e)
  }
}

async function extractAlbumImage(payload){
  if(!payload || !payload.strAlbumThumb){
    return
  }
  const url = payload.strAlbumThumb;
  const imageName = payload.idAlbum.concat('_album.jpg');

  return await extracted(url, imageName);
}

async function extractArtistImage(payload){
  if(!payload || !payload.strArtistThumb){
    return
  }

  const url = payload.strArtistThumb;
  const imageName = payload.idArtist.concat('_artist.jpg');

  return await extracted(url, imageName);
}

async function getOrCreateGenre(name){
//  checks if the genre exist, if so, it returns the genre object
//  else creates a new one with the name specified as parameter
  name = name.replace('/', '&')
  const genreColors = ["#5f0f40ff", "#7d092fff","#9a031eff","#cb4721ff", "#fb8b24ff",
    "#ef781cff", "#e36414ff","#ae5e26ff", "#795838ff", "#0f4c5cff"]

  let existingGenre = await Genre.findOne({where: {'name': name}})
  if (existingGenre === null){
    const randNum = Math.floor(Math.random() * genreColors.length);
    existingGenre = Genre.build({
      image: genreColors[randNum],
      name: name
    })
    await existingGenre.save();
  }
  return existingGenre;
}

async function validatePayload(payload) {

  if (payload.strArtist === undefined || payload.idArtist === undefined)
    return false;

  const existingArtist = await Artist.findByPk(payload.idArtist)
  if (existingArtist)
    return false;

  if (payload.genre === undefined)
    payload.genre = 'UNKNOWN';

  return true;
}

module.exports = { uploadArtists }